browsers/duckduckgo.safariextension/js/content-script.js

```diff
@@ -35,22 +29,36 @@ var onBeforeLoad = (e) => {
         if (block.cancel) {
             console.log(`DDG BLOCKING ${e.url}`)
             e.preventDefault()
+        } else if (block.redirectUrl) {
+            console.log(`DDG BLOCKING AND USING SURROGATE ${e.url}`)
+            setTimeout(redirectSrc.bind(undefined, e.target, block.redirectUrl), 1)
         }
     }
 
 }
 
-var unload = (e) => {
-    if (window === window.top) {
-        safari.self.tab.dispatchMessage('unloadTab', {unload: mainFrameURL})
-    }
+var onBeforeUnload = (e) => {
+    safari.self.tab.dispatchMessage('unloadTab', {
+        unload: getLocation()
+    })
 }
 
 // return location without params
+var mainFrameURL
 function getLocation () {
-    return location.protocol + "//" + location.hostname + location.pathname
+    if (mainFrameURL) { return mainFrameURL }
+    var loc = window.top.location
+    mainFrameURL = loc.protocol + "//" + loc.hostname + loc.pathname
+    return mainFrameURL
+}
+
+// serve surrogate content
+var redirectSrc = function(element, url) {
+    element.src = url
 }
 
-window.onbeforeunload = ((e) => unload(e))
+if (window === window.top) {
+    window.addEventListener('beforeunload', onBeforeUnload, true);
+}
```

Comment:
Unrelated, but I moved this to a listener so that there's no chance of sites overriding our handler here. Also only bound the event on window.top.

Reply:
Cool, good idea.